File: sum-of-multiples/sum_of_multiples.erl
-module(sum_of_multiples).
-export([sumOfMultiplesDefault/1, sumOfMultiples/2]).

sumOfMultiplesDefault(N) -> sumOfMultiples([3,5], N).

sumOfMultiples(Mult, N) -> lists:sum(anyDivides(Mult, lists:seq(1, N - 1))).

anyDivides(Factors, Numbers) ->
  lists:filter(fun(N) -> lists:any(fun(D) -> divides(D, N) end, Factors) end,
               Numbers).

divides(D, N) -> N rem D =:= 0.

File: sum-of-multiples/SumOfMultiples.swift
//
//  SumOfMultiples.swift
//  Exercism
//
//  Created by Eric Bailey on 6/17/16.
//
//

public struct SumOfMultiples {

  public static func toLimit(number: Int, inMultiples multiples: [Int]) -> Int {
    return Array(1..<number).anyDivides(multiples.filter { $0 != 0 }).sum()
  }

}

private extension Int {

  /** Return `true` iff `self` divides `dividend`.
   - parameter dividend: A number to divide by `self`
   - returns: `true` if `self` divides `dividend`, otherwise `false`.
   */
  func divides(dividend: Int) -> Bool {
    return dividend % self == 0
  }

}

private extension CollectionType where Generator.Element == Int {

  func anyDivides(factors: [Int]) -> [Int] {
    return filter { number in
      for factor in factors {
        if factor.divides(number) { return true }
      }

      return false
    }
  }

  func sum() -> Int {
    return reduce(0, combine: +)
  }

}
